### ISSUE TITLE
Error: Provider produced inconsistent result after apply

### ISSUE THREAD

**Issue by sidhukumar** [hashicorp/terraform-provider-azurerm#14344]
### Community Note

- Please vote on this issue by adding a 👍 reaction to the original issue to help the community and maintainers prioritize this request. 

- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request	

- If you are interested in working on this issue or have submitted a pull request, please leave a commen

`terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 2.86.0"
    }
    helm = {
      source = "hashicorp/helm"
      version = "1.3.2"
    }
    kubernetes = {
      source  = "hashicorp/kubernetes"
      version = "1.13.3"
    }
  }
}

provider "azurerm" {

  use_msi = var.azure_auth_method == "msi" ? true : false
  subscription_id = var.subscription_id
  client_id       = var.azure_auth_method == "principal" ? var.client_id : null
  client_secret   = var.azure_auth_method == "principal" ? var.client_secret : null
  tenant_id       = var.tenant_id

data "azurerm_resource_group" "starter_rg" {
  name     = var.resource_group_name
  #location = var.location
  #tags = local.tags
}

resource "azurerm_storage_account" "starter_storage_account" {
  name                      = var.storage_account_name
  enable_https_traffic_only = true
  resource_group_name       = data.azurerm_resource_group.starter_rg.name
  location                  = data.azurerm_resource_group.starter_rg.location
  account_tier              = var.storage_account_tier
  account_replication_type  = var.storage_account_replication_type
  is_hns_enabled            = true
  account_kind              = "StorageV2"
  access_tier               = var.storage_account_access_tier

  network_rules {
    default_action             = var.storage_account_firewall.default_action
    bypass                     = var.storage_account_firewall.bypass
    virtual_network_subnet_ids = [for sub_alias in var.storage_account_firewall.trusted_subnets: data.azurerm_subnet.subnets[sub_alias].id ]
    ip_rules                    = [for rule in var.storage_account_firewall.ip_rules: replace(rule, "myExtIP", data.http.myExtIP.body) ]
  }

  tags = local.tags

}

resource "azurerm_storage_account" "object_storage_4_datalake_storage_account" {
  name                      = var.blob_storage_account_name
  enable_https_traffic_only = true
  resource_group_name       = data.azurerm_resource_group.starter_rg.name
  location                  = data.azurerm_resource_group.starter_rg.location
  account_tier              = var.blob_storage_account_tier
  account_replication_type  = var.blob_storage_account_replication_type
  access_tier               = var.blob_storage_account_access_tier
  account_kind              = var.blob_storage_account_kind

  network_rules {
    default_action             = var.blob_storage_account_firewall.default_action
    virtual_network_subnet_ids = [for sub_alias in var.blob_storage_account_firewall.trusted_subnets: data.azurerm_subnet.subnets[sub_alias].id ]
    bypass                     = var.blob_storage_account_firewall.bypass
    ip_rules                    = [for rule in var.blob_storage_account_firewall.ip_rules: replace(rule, "myExtIP", data.http.myExtIP.body) ]
  }

  tags                      = local.tags
}

resource "azurerm_network_security_group" "databricks_network_sg" {
  count = var.databricks_custom_vnet_enabled == true ? 1 : 0
  name                = var.databricks_network_sg_name
  location            = data.azurerm_resource_group.starter_rg.location
  resource_group_name = data.azurerm_subnet.subnets[var.databricks_private_subnet].resource_group_name
}

resource "azurerm_subnet_network_security_group_association" "databricks_private_subnet_association" {
  count = var.databricks_custom_vnet_enabled == true ? 1 : 0
  subnet_id                 = data.azurerm_subnet.subnets[var.databricks_private_subnet].id
  network_security_group_id = azurerm_network_security_group.databricks_network_sg[0].id
}

resource "azurerm_subnet_network_security_group_association" "databricks_public_subnet_association" {
  count = var.databricks_custom_vnet_enabled == true ? 1 : 0
  subnet_id                 = data.azurerm_subnet.subnets[var.databricks_public_subnet].id
  network_security_group_id = azurerm_network_security_group.databricks_network_sg[0].id
}`

###  error mentioned below
The above is the line of codes that I have used and am facing an error while deploying the storage account and updating subnets by terraform using the azure DevOps pipeline.

### Debug Output

`Error: Provider produced inconsistent result after apply

When applying changes to azurerm_storage_account.starter_storage_account,
provider "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an
unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

Error: Provider produced inconsistent result after apply

When applying changes to azurerm_storage_account.ml_storage_account[0],
provider "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an
unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

Error: Provider produced inconsistent result after apply

When applying changes to
azurerm_storage_account.object_storage_4_datalake_storage_account, provider
"provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an
unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

Error: updating Network Security Group Association for Subnet: (Name "databricks-private-subnet" / Virtual Network Name "aiopsnet-aiac01" / Resource Group "aiaccelerator-vnet-demo"): network.SubnetsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidResourceReference" Message="Resource /subscriptions/b1239114-b332-4b01-94da-cd5159245abf/resourceGroups/AIACCELERATOR-VNET-DEMO/providers/Microsoft.Network/networkSecurityGroups/DATABRICKS-BP-MLAC100 referenced by resource /subscriptions/b1239114-b332-4b01-94da-cd5159245abf/resourceGroups/aiaccelerator-vnet-demo/providers/Microsoft.Network/virtualNetworks/aiopsnet-aiac01/subnets/databricks-private-subnet was not found. Please make sure that the referenced resource exists, and that both resources are in the same region." Details=[{"code":"NotFound","message":"Resource /subscriptions/b1239114-b332-4b01-94da-cd5159245abf/resourceGroups/AIACCELERATOR-VNET-DEMO/providers/Microsoft.Network/networkSecurityGroups/DATABRICKS-BP-MLAC100 not found."}]

  with azurerm_subnet_network_security_group_association.databricks_private_subnet_association[0],
  on batch-starter.tf line 871, in resource "azurerm_subnet_network_security_group_association" "databricks_private_subnet_association":
 871: resource "azurerm_subnet_network_security_group_association" "databricks_private_subnet_association" {

Error: updating Network Security Group Association for Subnet: (Name "databricks-public-subnet" / Virtual Network Name "aiopsnet-aiac01" / Resource Group "aiaccelerator-vnet-demo"): network.SubnetsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidResourceReference" Message="Resource /subscriptions/b1239114-b332-4b01-94da-cd5159245abf/resourceGroups/AIACCELERATOR-VNET-DEMO/providers/Microsoft.Network/networkSecurityGroups/DATABRICKS-BP-MLAC100 referenced by resource /subscriptions/b1239114-b332-4b01-94da-cd5159245abf/resourceGroups/aiaccelerator-vnet-demo/providers/Microsoft.Network/virtualNetworks/aiopsnet-aiac01/subnets/databricks-public-subnet was not found. Please make sure that the referenced resource exists, and that both resources are in the same region." Details=[{"code":"NotFound","message":"Resource /subscriptions/b1239114-b332-4b01-94da-cd5159245abf/resourceGroups/AIACCELERATOR-VNET-DEMO/providers/Microsoft.Network/networkSecurityGroups/DATABRICKS-BP-MLAC100 not found."}]

  with azurerm_subnet_network_security_group_association.databricks_public_subnet_association[0],
  on batch-starter.tf line 877, in resource "azurerm_subnet_network_security_group_association" "databricks_public_subnet_association":
 877: resource "azurerm_subnet_network_security_group_association" "databricks_public_subnet_association" {`

### Terraform (and AzureRM Provider) Version

provider azurerm verison = 2.86.0
terraform version 0.13.7

### Affected Resource(s)

Azure_storage_account
databricks_public_subnet
databricks_private_subnet

could you please help me with this issue?

### References
[10227](https://github.com/hashicorp/terraform-provider-azurerm/issues/10227)
[14298](https://github.com/hashicorp/terraform-provider-azurerm/issues/14298)

**Comment by jackofallops** [comment]
Hi @sidhukumar - Looking at the output, the service is suggesting that some of the resources you're referencing are either not valid for the way they are trying to be used, or simply don't exist rather than there being a bug in the provider.  For example:
> databricks-public-subnet was not found. Please make sure that the referenced resource exists, and that both resources are in the same region.

**Comment by sidhukumar** [comment]
No, I can see the NSG is been deployed in the resource group and I have also checked the regions they are deployed in west us 
and the subnet is also pointing out to the west us. i have attached screeshot for reference.

![image](https://user-images.githubusercontent.com/31894412/143441576-b566c20d-7eea-476d-b6a4-17e617450f7f.png)

![image](https://user-images.githubusercontent.com/31894412/143442532-d445ba8d-96c4-4a05-8a23-7d2d697d954a.png)


the bug in the providers is all pointing out to the storage account. I have tried using older versions. I still get the same error.

 `Error: Provider produced inconsistent result after apply

When applying changes to azurerm_storage_account.ml_storage_account[0],
provider "provider["registry.terraform.io/hashicorp/azurerm"]" produced an
unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.`
